src/StackExchange.Redis/Configuration/LoggingTunnel.cs

```diff
@@ -0,0 +1,624 @@
+﻿using Pipelines.Sockets.Unofficial;
+using Pipelines.Sockets.Unofficial.Arenas;
+using System;
+using System.Buffers;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.IO;
+using System.IO.Pipelines;
+using System.Net;
+using System.Net.Security;
+using System.Net.Sockets;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using static StackExchange.Redis.PhysicalConnection;
+
+namespace StackExchange.Redis.Configuration;
+
+/// <summary>
+/// Captures redis traffic; intended for debug use 
+/// </summary>
+[Obsolete("This API is experimental, has security and performance implications, and may change without notice", false)]
+[SuppressMessage("ApiDesign", "RS0016:Add public types and members to the declared API", Justification = "Experimental API")]
+public abstract class LoggingTunnel : Tunnel
+{
+    private readonly ConfigurationOptions _options;
+    private readonly bool _ssl;
+    private readonly Tunnel? _tail;
+
+    /// <summary>
+    /// Replay the RESP messages for a pair of streams, invoking a callback per operation
+    /// </summary>
+    public static async Task<long> ReplayAsync(Stream @out, Stream @in, Action<RedisResult, RedisResult> pair)
+    {
+        using Arena<RawResult> arena = new();
+        var outPipe = StreamConnection.GetReader(@out);
+        var inPipe = StreamConnection.GetReader(@in);
+
+        long count = 0;
+        while (true)
+        {
+            var sent = await ReadOneAsync(outPipe, arena, isInbound: false);
```

Comment:
Append `.ForAwait();` on async calls?

Reply:
good spot; fixed in 64f31748b70998614a0ea06f5c0d07743540254c, although oddly that isn't showing on GitHub yet (despite being pushed)